File: Design-Patterns-Using-Java/Decorator/DecoratorPatternWindow.java
import java.util.ArrayList;
import java.util.List;

// Window (Component)
interface Window {
    void draw();
}

// SimpleWindow (ConcreteComponent)
class SimpleWindow implements Window {
    @Override
    public void draw() {
        System.out.println("SimpleWindow draw.");
    }
}

// Decorator (Decorator)
abstract class Decorator implements Window {
    protected Window component;

    public Decorator(Window component) {
        this.component = component;
    }

    @Override
    public void draw() {
        component.draw();
    }
}

// VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator extends Decorator {
    public VerticalScrollBarDecorator(Window component) {
        super(component);
    }

    @Override
    public void draw() {
        super.draw();
        System.out.println("VerticalScrollBarDecorator draw");
    }
}

// HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator extends Decorator {
    public HorizontalScrollBarDecorator(Window component) {
        super(component);
    }

    @Override
    public void draw() {
        super.draw();
        System.out.println("HorizontalScrollBarDecorator draw");
    }
}

// Client code
public class DecoratorPatternWindow {
    public static void main(String[] args) {
        Window component = new SimpleWindow();
        Window decorator1 = new VerticalScrollBarDecorator(component);
        Window decorator2 = new HorizontalScrollBarDecorator(decorator1);
        decorator2.draw();
    }
}
/*
SimpleWindow draw.
VerticalScrollBarDecorator draw
HorizontalScrollBarDecorator draw
 */

File: Design-Patterns-Using-Kotlin/Decorator/DecoratorPatternWindow.kt
// Window (Component)
interface Window {
    fun draw()
}

// SimpleWindow (ConcreteComponent)
class SimpleWindow : Window {
    override fun draw() {
        println("SimpleWindow draw.")
    }
}

// Decorator (Decorator)
abstract class Decorator(protected var component: Window) : Window {
    override fun draw() {
        component.draw()
    }
}

// VerticalScrollBarDecorator (ConcreteDecorator)
class VerticalScrollBarDecorator(component: Window) : Decorator(component) {
    override fun draw() {
        super.draw()
        println("VerticalScrollBarDecorator draw")
    }
}

// HorizontalScrollBarDecorator (ConcreteDecorator)
class HorizontalScrollBarDecorator(component: Window) : Decorator(component) {
    override fun draw() {
        super.draw()
        println("HorizontalScrollBarDecorator draw")
    }
}

// Client code
fun main() {
    val component: Window = SimpleWindow()
    val decorator1: Window = VerticalScrollBarDecorator(component)
    val decorator2: Window = HorizontalScrollBarDecorator(decorator1)
    decorator2.draw()
}

/*
SimpleWindow draw.
VerticalScrollBarDecorator draw
HorizontalScrollBarDecorator draw
*/
